themes/default/content/learn/abstraction-encapsulation/abstraction/index.md

```diff
@@ -0,0 +1,68 @@
+---
+title: "Abstracting Pulumi Code"
+layout: topic
+date: 2021-11-17
+draft: false
+description: |
+    Learn more about abstracting Pulumi code into classes, models, and objects.
+meta_desc: |
+    Learn more about abstracting Pulumi code into classes, models, and objects.
+index: 0
+estimated_time: 5
+meta_image: meta.png
+authors:
+    - laura-santamaria
+tags:
+    - learn
+block_external_search_index: false
+---
+
+Just like any application code, Pulumi infrastructure code can be abstracted,
+enabling us to work with models or objects, depending on which language we use
+to define that code. If we use an object-oriented langauge such as Javascript,
+Python, or Typescript, we can create and instantiate classes. For languages like
+Go, we can build up interfaces. Either way, we're thinking in terms of taking
+code that we've written and making it reusable in some form.
+
+Breaking apart Pulumi code is essential to keeping a clean, well-organized, and
+easily maintained code base that enables teams to collaborate to enhance current
+systems. As the owners of the infrastructure for the Pulumipus brand, we've got
+a lot of resources to maintain, and it's much better to reuse code that we write
+once rather than repeating ourselves continuously in our program and throughout
+all of the programs we maintain.
+
+Let's first explore abstraction.
+
+## Thinking in abstraction
+
+When we use our code for our Pulumi program, we're generally starting out with a
+declarative thought process. We write code that declares we want three resources
+of a specific type, in that order, and run the code to create those three
+resources in order.
+
+To get to more complex architectures, we need to enable the ability to break
+apart that code so we can reuse components and call things in any order we
+choose, knowing that we will get the expected result each time. We also need to
+allow others to use parts of our code without necessarily needing to know
+precisely how it's put together. This abstractionist way of thinking means we're
+thinking in terms of the mental models we use to understand the cloud
+architectures that are available to us&mdash;or what we want to make available
+to others in our organization.
+
+## Exploring an example
+
+Let's imagine the Pulumipus brand is expanding to all kinds of systems, and so
+we don't know which infrastructure each team will need. However, each team will
+need some kind of storage location that has an access policy, no matter whether
+they're doing a serverless build or a compute-based build.
+
+Mentally, we're defining a sort of abstraction like this:
+
+![A diagram showing a user doing an action that touches the storage through a policy](./abstraction.png)
+
+Seems pretty clean, doesn't it? Abstraction is thinking through what we
```

Comment:
To be honest, I don't understand what this diagram is telling me. Are "user" and "action" _part_ of the abstraction somehow? Or are they stand-ins for the consumers of the `policy + storage` abstraction that we develop in the pages that follow? I also feel, as I reach the end of this page, like I still don't quite grok what "abstraction" means in this context, or what problem it solves; the words all make sense, and the points you've made all ring true, but somehow, I still seem to be missing the take-away.

One thing that might help would be to consider explaining encapsulation first, since that's often what you reach for before you start thinking about higher-level abstractions. When you realize you've copy-pasted the same block of code to provision a bucket and policy a dozen times, it's sometimes enough just wrap it up into a function -- `makeBucket("my-bucket-name")`, say -- to reduce all that duplication. That qualifies as encapsulation -- tucking away the repetition and exposing only the parts that need to vary -- but not really abstraction, since the concept of "bucket" still exists in the mind of the person calling the `makeBucket()` API.

From there, you could introduce abstraction to solve the problem of APIs that are too low-level. If the reason your users are creating all these buckets is to serve static websites, for example, and those websites also require provisioning additional things like domain names or other stuff, it'd probably might be much just nicer to give them a higher-level API, like `new Website("something.com")`, than to expect them to call lots of functions to create buckets, A records, TLS certs, and so on.

Does this make sense? And then finally, bring in components to solve the problem of conveying how collections of resources are related to one another -- how all these dozens of policies and things all ultimately fall under the `Website` component, for example. Anyway, let me know what you think -- I'm not married to any of this or anything, I'm just thinking of ways to help bring these points home is all.

Reply:
No, this makes sense! I'm used to thinking in broad strokes and then drilling down, but that's how my brain works since I'm often trying to find patterns so I can pull out more accessible analogies (e.g., talking about state from physics to explain state in infrastructure) to explain the concepts to more people. I should have thought more about that flip when putting this together. That will also make the diagram make more sense (I was trying to diagram the S3 bucket example in the most generic form possible; it probably didn't help that I didn't use a standard architecture program to make the diagram 🤔  ). Let me rework this.